### ISSUE TITLE
Rename Action to Callback and Callback to JSCallback

### ISSUE THREAD

**Issue by birdsarah** [bokeh/bokeh#2634]
At PyData London, @bryevdv and I got into this weird disagreement about a new callback. After a while I realized that I was just confused by the current naming structure.

Right now in https://github.com/bokeh/bokeh/blob/master/bokeh/models/actions.py

we have

```
- Action(PlotObject)
-- Callback(Action)
-- OpenURL(Action)
```

In most of the codebase where we have callbacks, we have

`callback = Instance(Callback, ....)`

but on TapTool we have 

`action = Instance(Action,...)`

with the example centered around OpenURL. Of course, action could actually be a Callback, as Callback is also a subclass of Action.

This inconsistency is just a historical accident as OpenURL came before Callback.

I would like to propose the following, backwards incompatible, changes.

1 - rename `Action` to `Callback` and `Callback` to `CustomJS`
```
- Callback(PlotObject)
-- OpenURL(Callback)
-- CustomJS(Callback)
```

2 - replace TapTool `action` kwarg to be `callback`.

All models with callbacks have the property as previously:

`callback = Instance(Callback, ....)`

(of course, this is now referring to the higher level class).

Given that we have been talking about new "`JupyterCallback`"s and "`ServerCallback`"s I imagine that these would also subclass Callback and be acceptable inputs to a model that accepts a callback.  So for example you could imagine all of these as possible future definitions of a TapTool

```python
customjs_tap = TapTool(callback=CustomJS(args={}, code="alert('I was tapped');"))
openurl_tap = TapTool(callback=OpenURL(url="http://openontap.com"))
jupyter_tap = TapTool(callback=Jupyter(args={}, method=lambda: print('i was tapped')))
```

Given that `Callback` and even `Action` are pretty new, I think this is the best time to do it before we start adding a lot more canned "callbacks."

**Comment by canavandl** [comment]
I agree w/ the proposed naming structure. I would think that it's more semantically correct.

I can't comment on the cost/benefit of making a backwards incompatible change, but I guess the callbacks are a pretty new addition to Bokeh, so it might not be that bad.

**Comment by birdsarah** [comment]
In the end I made deprecation functions so it shouldn't break anyone's existing code, with DeprecationWarnings for v0.10.

**Comment by jakirkham** [comment]
So, I ran into an issue related to this today. However, I am very happy to see that you have explained your rational and added this here. It was easy enough to fix the issue from the information here. Also, I am glad to see you went to the effort to add a warning.

That all being said, in iPython I think the warning is being swallowed. Instead, I get a rather nasty exception. Here is a reproducer of the problem I am seeing ( https://gist.github.com/jakirkham/dfa04a13d5aa694c504b ).

This probably should go in a new issue, but I am putting it here for now.

**Comment by damianavila** [comment]
> This probably should go in a new issue, but I am putting it here for now.

Yes, please... thanks!

**Comment by jakirkham** [comment]
Done. ( https://github.com/bokeh/bokeh/issues/2791 )
